README.md
# Logging support code for **Stakker**

[![license:MIT/Apache-2.0][1]](https://github.com/uazu/stakker_log)
[![github:uazu/stakker_log][2]](https://github.com/uazu/stakker_log)
[![crates.io:stakker_log][3]](https://crates.io/crates/stakker_log)
[![docs.rs:stakker_log][4]](https://docs.rs/stakker_log)

[1]: https://img.shields.io/badge/license-MIT%2FApache--2.0-blue
[2]: https://img.shields.io/badge/github-uazu%2Fstakker_log-brightgreen
[3]: https://img.shields.io/badge/crates.io-stakker_log-red
[4]: https://img.shields.io/badge/docs.rs-stakker_log-purple

This provides logging macros, logging type conversions and other
logging support code for
[**Stakker**](https://crates.io/crates/stakker).  It is planned to add
logging forwarders for common logging crates.

### Documentation

See the [crate documentation](http://docs.rs/stakker_log).

# License

This project is licensed under either the Apache License version 2 or
the MIT license, at your option.  (See
[LICENSE-APACHE](LICENSE-APACHE) and [LICENSE-MIT](LICENSE-MIT)).

### Contribution

Unless you explicitly state otherwise, any contribution intentionally
submitted for inclusion in this crate by you, as defined in the
Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Significant feature changes and additions

<!-- see keepachangelog.com for format ideas -->

## 0.1.0 (2022-06-30)

### Added

- JSON formatting of log data, using `KvToJson` type

## 0.0.3 (2021-01-26)

### Added

- Add `LogCx`
- Make `Option` visitable
- Allow tag expr for `audit!`

<!-- Local Variables: -->
<!-- mode: markdown -->
<!-- End: -->


src/test.rs
use stakker::*;
use std::cell::Cell;
use std::collections::HashMap;
use std::net::Ipv4Addr;
use std::rc::Rc;
use std::time::Instant;

use crate::{error, KvSingleLine, Visitable};

// TODO: Need tests of all the different shortcuts
// TODO: Need test of audit!

struct MyType;
impl Visitable for MyType {
    fn visit(&self, key: Option<&str>, out: &mut dyn LogVisitor) {
        out.kv_map(key);
        out.kv_u64(Some("a"), 135);
        out.kv_null(Some("b"));
        out.kv_arr(Some("c"));
        out.kv_arrend(Some("c"));
        out.kv_mapend(key);
    }
}

#[test]
fn error_formatting() {
    let mut stakker = Stakker::new(Instant::now());
    let s = &mut stakker;
    let out = Rc::new(Cell::new(String::new()));
    let out2 = out.clone();

    s.set_logger(
        LogFilter::all(&[LogLevel::Trace, LogLevel::Audit, LogLevel::Open]),
        move |_, r| {
            out2.set(format!(
                "{} #{} {} {}",
                r.level,
                r.id,
                r.fmt,
                KvSingleLine::new(r.kvscan, "{", "}")
            ));
        },
    );

    let a = "TEST";
    let b = 1.234_f64;
    let c = 1234_i32;
    let d = -1234_i16;
    let e = 4321_u64;
    let f = vec!["abc", "def"];
    let g = Ipv4Addr::LOCALHOST;
    let h = ();
    let mut i = HashMap::new();
    i.insert("a", "cat");
    i.insert("b", "dog");
    let j = "This is a test";
    let k = MyType;
    error!([s], a, b, c, d, e, f, %g, h, i, j, k, "Test");
    let o = out.take();
    // Hashmap is unordered, so there are two possibilities
    match o.as_str() {
        "ERROR #0 Test {a=TEST b=1.234 c=1234 d=-1234 e=4321 f[abc def] \
         g=127.0.0.1 h i{a=cat b=dog} j=\"This is a test\" k{a=135 b c[]}}" => (),
        "ERROR #0 Test {a=TEST b=1.234 c=1234 d=-1234 e=4321 f[abc def] \
         g=127.0.0.1 h i{b=dog a=cat} j=\"This is a test\" k{a=135 b c[]}}" => (),
        _ => panic!("Unexpected output: {}", o),
    }
}


src/lib.rs
//! Logging macros for **Stakker**
//!
//! There are five severity-based logging macros ([`trace!`],
//! [`debug!`], [`info!`], [`warn!`] and [`error!`]) and one macro
//! designed for logging records that have a fixed tag and no freeform
//! text ([`audit!`]).  Examples:
//!
//! ```ignore
//! error!([cx], addr: %src_addr, port, "Failed to connect: {}", err);
//! audit!([cx], TcpConnectFailure, addr: %src_addr, port);
//! ```
//!
//! The severity-based macros like [`error!`] all work the same way.
//! The `[cx]` comes first, followed by an optional target
//! specification (`target: "target-name"`), followed by optional
//! key-value pairs, followed by a format-string and its arguments.
//!
//! For [`audit!`], `[cx]` comes first, followed by a tag for the
//! record, followed by key-value pairs.  The tag will normally be a
//! plain identifier, but it could also be a literal string or an
//! expression in parentheses which will be formatted to generate the
//! tag.
//!
//! `[cx]` can refer to either an actor context (`stakker::Cx`) or a
//! [`LogCx`].  Where the call is not being made from a context that
//! provides a `LogID`, `[core]` may be passed instead of `[cx]`,
//! which gives a `LogID` of zero.  It's possible to log against a
//! specific actor or other `LogID` source by using `[source, core]`
//! instead of `[cx]`, which takes the `LogID` from that source using
//! a `source.access_log_id()` call.  (In general the `[a]` form must
//! support `a.access_log_id()` and `a.access_core()`, and the `[a,b]`
//! form must support `a.access_log_id()` and `b.access_core()`.)
//!
//! For key-value pairs, the most general form is `"key": expr`, but
//! there are a number of shortcuts as follows:
//!
//! Shortcut | Equivalent
//! --- | ---
//! `size: file_size` | `"size": file_size`
//! `size` | `"size": size`
//! `packet.size` | `"size": packet.size`
//! `tcp.packet.size` | `"size": tcp.packet.size`
//! `%src_addr` | `"src_addr": format_args!("{}", src_addr)`
//! `src_addr: %addr` | `"src_addr": format_args!("{}", addr)`
//! `?stream` | `"stream": format_args!("{:?}", stream)`
//! `stream: ?input_stream` | `"stream": format_args!("{:?}", input_stream)`
//!
//! Conversion of values is determined by implementation of the
//! [`Visitable`] trait.  All Rust primitives and standard collections
//! are supported by [`Visitable`].  For your own types, there are two
//! possibilities: either implement `Display` or `Debug` and use `%`
//! or `?`, or else implement [`Visitable`] directly.  The advantage
//! of implementing [`Visitable`] is that in addition to mapping to a
//! primitive instead of a string, you could also output a structured
//! value such as an array or map to represent your type.
//!
//! # Logging output
//!
//! You can write you own code which accepts a `&dyn Fn(&mut dyn
//! LogVisitor)`, and calls it to receive all the logging data.  There
//! are also provided types for JSON output ([`KvToJson`]) and simple
//! human-readable output ([`KvSingleLine`]).
//!
//! [`KvSingleLine`]: struct.KvSingleLine.html
//! [`KvToJson`]: struct.KvToJson.html
//! [`LogCx`]: struct.LogCx.html
//! [`Visitable`]: trait.Visitable.html
//! [`audit!`]: macro.audit.html
//! [`debug!`]: macro.debug.html
//! [`error!`]: macro.error.html
//! [`info!`]: macro.info.html
//! [`trace!`]: macro.trace.html
//! [`warn!`]: macro.warn.html

mod kvdisp;
mod kvjson;
mod logcx;
mod macros;
mod visit;

pub use kvdisp::KvSingleLine;
pub use kvjson::KvToJson;
pub use logcx::LogCx;
pub use visit::Visitable;

// Re-export so that macros can access stakker::LogLevel
#[doc(hidden)]
pub use stakker;

// TODO: Add loggers that log to 'log' crate, 'slog' crate,
// 'tracing-core', etc

#[cfg(test)]
mod test;


src/logcx.rs
use stakker::{Core, LogID};

/// Logging context
///
/// This encapsulates a [`stakker::LogID`] and a reference to
/// [`stakker::Core`].  It allows logging to a particular context or
/// span without having to carry around a reference to a particular
/// actor.  A reference to a [`LogCx`] can be used as the `[cx]`
/// argument to any of the logging macros.
///
/// [`LogCx`]: struct.LogCx.html
/// [`stakker::Core`]: ../stakker/struct.Core.html
/// [`stakker::LogID`]: ../stakker/type.LogID.html
pub struct LogCx<'a> {
    logid: LogID,
    core: &'a mut Core,
}

impl<'a> LogCx<'a> {
    /// Create directly from `LogID` and `Core` reference
    pub fn new(logid: LogID, core: &'a mut Core) -> Self {
        Self { logid, core }
    }

    /// Used by macros to obtain the `LogID`
    pub fn access_log_id(&self) -> LogID {
        self.logid
    }

    /// Used by macros to obtain the `Core` reference
    pub fn access_core(&mut self) -> &mut Core {
        self.core
    }
}


src/visit.rs
use stakker::LogVisitor;
use std::collections::{BTreeMap, BTreeSet, BinaryHeap, HashMap, HashSet, LinkedList, VecDeque};
use std::fmt::Arguments;

/// This trait allows a `stakker::LogVisitor` to visit various
/// fundamental Rust types and collections.
///
/// This allows all of those types to be used as values in the logging
/// macros.  It does this by mapping them to the fixed set of `kv_*`
/// methods available on the `stakker::LogVisitor`.  Most fundamental
/// Rust types, plus map-like and array-like collections have a
/// straightforward mapping.  Integer types larger than 64-bits are
/// output as strings.  `Option` is handled by including the key-value
/// pair for `Some`, or omitting it for `None`.
///
/// For other types which must be formatted as a string, construct a
/// `std::fmt::Arguments` instance using `format_args!` first, and
/// then visit that.  For your own complex types which need structured
/// output, you should write your own [`Visitable`] implementation.
///
/// For example:
///
/// ```notest
/// var1.visit("key1", visitor);     // Handle according to type
/// var2.visit("key2", visitor);
/// format_args!("{}", var3).visit("key3", visitor);   // Display formatting
/// format_args!("{:?}", var4).visit("key4", visitor); // Debug formatting
/// ```
///
/// [`Visitable`]: trait.Visitable.html
pub trait Visitable {
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor);
}

// Option handling
impl<T: Visitable> Visitable for Option<T> {
    #[inline]
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
        if let Some(ref v) = self {
            v.visit(key, output);
        }
    }
}

// String handling
impl Visitable for &str {
    #[inline]
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
        output.kv_str(key, *self);
    }
}

impl Visitable for String {
    #[inline]
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
        output.kv_str(key, self.as_str());
    }
}

// Null or unit handling
impl Visitable for () {
    #[inline]
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
        output.kv_null(key);
    }
}

// Format arguments handling
impl<'a> Visitable for Arguments<'a> {
    #[inline]
    fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
        output.kv_fmt(key, self);
    }
}

// Copy types and convertible types
macro_rules! visit_copy_as {
    ($fr:ty, $to:ty, $method:ident) => {
        impl $crate::Visitable for $fr {
            #[inline]
            fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
                output.$method(key, *self as $to);
            }
        }
    };
}

visit_copy_as!(u8, u64, kv_u64);
visit_copy_as!(u16, u64, kv_u64);
visit_copy_as!(u32, u64, kv_u64);
visit_copy_as!(u64, u64, kv_u64);
visit_copy_as!(usize, u64, kv_u64);
visit_copy_as!(i8, i64, kv_i64);
visit_copy_as!(i16, i64, kv_i64);
visit_copy_as!(i32, i64, kv_i64);
visit_copy_as!(i64, i64, kv_i64);
visit_copy_as!(isize, i64, kv_i64);
visit_copy_as!(f32, f64, kv_f64);
visit_copy_as!(f64, f64, kv_f64);
visit_copy_as!(bool, bool, kv_bool);

// Types that we have to just format out as a string
macro_rules! visit_as_display {
    ($fr:ty) => {
        impl $crate::Visitable for $fr {
            #[inline]
            fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
                output.kv_fmt(key, &format_args!("{}", *self));
            }
        }
    };
}

visit_as_display!(char);
visit_as_display!(u128);
visit_as_display!(i128);

// Array-like objects
macro_rules! visit_arr {
    ($t:ident, $fr:ty) => {
        impl<$t: Visitable> Visitable for $fr {
            #[inline]
            fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
                output.kv_arr(key);
                for v in self.iter() {
                    v.visit(None, output);
                }
                output.kv_arrend(key);
            }
        }
    };
}

visit_arr!(T, [T]);
visit_arr!(T, Vec<T>);
visit_arr!(T, VecDeque<T>);
visit_arr!(T, LinkedList<T>);
visit_arr!(T, HashSet<T>);
visit_arr!(T, BTreeSet<T>);
visit_arr!(T, BinaryHeap<T>);

// Map-like objects
macro_rules! visit_map {
    ($fr:ident) => {
        impl<K: AsRef<str>, V: Visitable> Visitable for $fr<K, V> {
            #[inline]
            fn visit(&self, key: Option<&str>, output: &mut dyn LogVisitor) {
                output.kv_map(key);
                for (k, v) in self {
                    v.visit(Some(k.as_ref()), output);
                }
                output.kv_mapend(key);
            }
        }
    };
}

visit_map!(HashMap);
visit_map!(BTreeMap);


src/kvdisp.rs
use stakker::LogVisitor;
use std::fmt;
use std::fmt::Arguments;
use std::fmt::Write;

/// Single-line rendering of key-value pairs
///
/// When formatted with `"{}"`, this produces a human-readable
/// single-line rendering of the key-value pairs.  It is good for
/// debugging or simple display, or for input into human-readable
/// string-based processing.  All type information on key values is
/// lost.  Strings are shown without quotes if no characters need
/// quoting.  Simple `\XX` escaping is used for reserved ASCII
/// characters, where `XX` is two hex digits.  Anything higher than
/// ASCII is passed unchanged.  Arrays are enclosed in `[...]` and
/// maps are enclosed in `{...}`.
pub struct KvSingleLine<'a> {
    kvscan: &'a dyn Fn(&mut dyn LogVisitor),
    prefix: &'static str,
    suffix: &'static str,
}

impl<'a> KvSingleLine<'a> {
    /// Create a `KvSingleLine` ready to be formatted.  `prefix` and
    /// `suffix` are two strings which are output before and after the
    /// key-value pairs, but only if the list of key-value pairs is
    /// non-empty.
    pub fn new(
        kvscan: &'a dyn Fn(&mut dyn LogVisitor),
        prefix: &'static str,
        suffix: &'static str,
    ) -> Self {
        Self {
            kvscan,
            prefix,
            suffix,
        }
    }
}

impl<'a> fmt::Display for KvSingleLine<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut visitor = Visitor::new(f, self.prefix);
        (self.kvscan)(&mut visitor);
        if visitor.error {
            Err(fmt::Error)
        } else if visitor.empty {
            Ok(()) // Didn't output anything
        } else {
            f.write_str(self.suffix)
        }
    }
}

// Reserved characters outside quotes
#[inline]
fn is_reserved(ch: char) -> bool {
    ch <= ' '
        || ch == '"'
        || ch == '='
        || ch == '\\'
        || ch == '['
        || ch == ']'
        || ch == '{'
        || ch == '}'
}

// This has to be outside Visitor due to borrowing issues
#[inline]
fn push_str_val(f: &mut fmt::Formatter<'_>, val: &str) -> fmt::Result {
    if val.find(is_reserved).is_some() {
        f.write_char('"')?;
        for ch in val.chars() {
            if ch < ' ' || ch == '"' || ch == '\\' {
                write!(f, "\\{:02X}", ch as u8)?;
            } else {
                f.write_char(ch)?;
            }
        }
        f.write_char('"')?;
    } else {
        f.write_str(val)?;
    }
    Ok(())
}

// Catch error return and set error flag
macro_rules! catch {
    ($self:ident, $call:expr) => {{
        if $call.is_err() {
            $self.error = true;
        }
    }};
}

struct Visitor<'a, 'b: 'a> {
    fmt: &'a mut fmt::Formatter<'b>,
    fmtbuf: String,
    prefix: &'static str, // Whatever needs adding before the next item, or ""
    empty: bool,
    error: bool,
}

impl<'a, 'b> Visitor<'a, 'b> {
    fn new(fmt: &'a mut fmt::Formatter<'b>, prefix: &'static str) -> Self {
        Self {
            fmt,
            fmtbuf: String::new(),
            prefix,
            empty: true,
            error: false,
        }
    }
    fn push_key(&mut self, key: Option<&str>, sep: Option<char>) {
        catch!(self, self.fmt.write_str(self.prefix));
        self.empty = false;
        self.prefix = " ";
        if let Some(key) = key {
            if key.is_empty() {
                catch!(self, self.fmt.write_str("\\20"));
            } else {
                for ch in key.chars() {
                    if is_reserved(ch) {
                        catch!(self, write!(self.fmt, "\\{:02X}", ch as u8));
                    } else {
                        catch!(self, self.fmt.write_char(ch));
                    }
                }
            }
            if let Some(sep) = sep {
                catch!(self, self.fmt.write_char(sep));
            }
        }
    }
}

impl<'a, 'b> LogVisitor for Visitor<'a, 'b> {
    fn kv_u64(&mut self, key: Option<&str>, val: u64) {
        self.push_key(key, Some('='));
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_i64(&mut self, key: Option<&str>, val: i64) {
        self.push_key(key, Some('='));
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_f64(&mut self, key: Option<&str>, val: f64) {
        self.push_key(key, Some('='));
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_bool(&mut self, key: Option<&str>, val: bool) {
        self.push_key(key, Some('='));
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_null(&mut self, key: Option<&str>) {
        self.push_key(key, None);
    }
    fn kv_str(&mut self, key: Option<&str>, val: &str) {
        self.push_key(key, Some('='));
        catch!(self, push_str_val(self.fmt, val));
    }
    fn kv_fmt(&mut self, key: Option<&str>, val: &Arguments<'_>) {
        self.push_key(key, Some('='));
        if self.fmtbuf.capacity() == 0 {
            self.fmtbuf = String::with_capacity(1024);
        }
        self.fmtbuf.clear();
        catch!(self, write!(self.fmtbuf, "{}", val));
        catch!(self, push_str_val(self.fmt, &self.fmtbuf));
    }
    fn kv_map(&mut self, key: Option<&str>) {
        self.push_key(key, None);
        catch!(self, self.fmt.write_str("{"));
        self.prefix = "";
    }
    fn kv_mapend(&mut self, _: Option<&str>) {
        catch!(self, self.fmt.write_str("}"));
        self.prefix = " ";
    }
    fn kv_arr(&mut self, key: Option<&str>) {
        self.push_key(key, None);
        catch!(self, self.fmt.write_str("["));
        self.prefix = "";
    }
    fn kv_arrend(&mut self, _: Option<&str>) {
        catch!(self, self.fmt.write_str("]"));
        self.prefix = " ";
    }
}

#[cfg(test)]
mod test {
    use super::{KvSingleLine, LogVisitor};
    use std::fmt::Write;

    fn kvscan(lv: &mut dyn LogVisitor) {
        lv.kv_u64(Some("u64"), 123456789);
        lv.kv_i64(Some("i64"), -123456789);
        lv.kv_f64(Some("f64"), 12345.6789);
        lv.kv_bool(Some("b0"), false);
        lv.kv_bool(Some("b1"), true);
        lv.kv_null(Some("null"));
        lv.kv_str(Some("str"), "ABCDEFGHIJ");
        lv.kv_str(Some("str_ctrl"), "ABC\tDEF");
        lv.kv_str(Some("str_quote"), "ABC\"DEF\"GHI");
        lv.kv_str(Some("str_bsl"), "ABC\\DEF\\GHI");
        lv.kv_fmt(Some("fmt"), &format_args!("{}{}{}", "ABC", 123, "DEF"));
        lv.kv_map(Some("map"));
        lv.kv_u64(Some("map_u64"), 987654321);
        lv.kv_str(Some("map_str"), "JIHGFEDCBA");
        lv.kv_map(Some("map_nested"));
        lv.kv_bool(Some("map_nested_bool"), false);
        lv.kv_mapend(Some("map_nested"));
        lv.kv_mapend(Some("map"));
        lv.kv_map(Some("map_empty"));
        lv.kv_mapend(Some("map_empty"));
        lv.kv_arr(Some("arr"));
        lv.kv_u64(None, 987654321);
        lv.kv_str(None, "JIHGFEDCBA");
        lv.kv_arr(None);
        lv.kv_bool(None, true);
        lv.kv_arrend(None);
        lv.kv_arrend(Some("arr"));
        lv.kv_arr(Some("arr_empty"));
        lv.kv_arrend(Some("arr_empty"));
    }

    fn append(
        s: &mut String,
        kvscan: &dyn Fn(&mut dyn LogVisitor),
        prefix: &'static str,
        suffix: &'static str,
    ) {
        write!(s, "{}", KvSingleLine::new(kvscan, prefix, suffix)).unwrap();
    }

    /// Basic sanity-check
    #[test]
    fn test() {
        let mut buf = "dummy=1".to_string();
        append(&mut buf, &kvscan, " ", "");
        println!("{}", buf);
        assert_eq!(buf, "dummy=1 u64=123456789 i64=-123456789 f64=12345.6789 b0=false b1=true null str=ABCDEFGHIJ str_ctrl=\"ABC\\09DEF\" str_quote=\"ABC\\22DEF\\22GHI\" str_bsl=\"ABC\\5CDEF\\5CGHI\" fmt=ABC123DEF map{map_u64=987654321 map_str=JIHGFEDCBA map_nested{map_nested_bool=false}} map_empty{} arr[987654321 JIHGFEDCBA [true]] arr_empty[]");
    }
}


src/kvjson.rs
use stakker::LogVisitor;
use std::fmt;
use std::fmt::Arguments;
use std::fmt::Write;

/// JSON rendering of key-value pairs
///
/// When formatted with `"{}"`, this produces a single-line compact
/// JSON rendering of the key-value pairs.
pub struct KvToJson<'a> {
    kvscan: &'a dyn Fn(&mut dyn LogVisitor),
    prefix: &'static str,
    suffix: &'static str,
}

impl<'a> KvToJson<'a> {
    /// Create a `KvToJson` ready to be formatted.  The output will
    /// have commas between items, but not at the start or end.  A
    /// list of key-value pairs would normally be surrounded by `{`
    /// and `}` in JSON, but since this type may be used in other
    /// contexts, e.g. to extend a JSON object, only the plain
    /// key-value pairs are output.
    ///
    /// `prefix` and `suffix` are two strings which are output before
    /// and after the key-value pairs, but only if the list of
    /// key-value pairs is non-empty.  So, for example, passing `","`
    /// and `""` would be suitable for extending a larger JSON object
    /// that you're already building up.  Or alternatively, passing
    /// `",\"kv\":{"` and `"}"` would allow an optional `"kv"` item to
    /// be added to a larger JSON object, but only if there is
    /// key-value data.
    pub fn new(
        kvscan: &'a dyn Fn(&mut dyn LogVisitor),
        prefix: &'static str,
        suffix: &'static str,
    ) -> Self {
        Self {
            kvscan,
            prefix,
            suffix,
        }
    }
}

impl<'a> fmt::Display for KvToJson<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut visitor = Visitor::new(f, self.prefix);
        (self.kvscan)(&mut visitor);
        if visitor.error {
            Err(fmt::Error)
        } else if visitor.empty {
            Ok(()) // Didn't output anything
        } else {
            f.write_str(self.suffix)
        }
    }
}

/// JSON string quoting
#[inline]
fn push_str_literal(f: &mut fmt::Formatter<'_>, val: &str) -> fmt::Result {
    f.write_char('"')?;
    if val.find(|ch| ch < ' ' || ch == '"' || ch == '\\').is_some() {
        for ch in val.chars() {
            match ch {
                '"' | '\\' => {
                    f.write_char('\\')?;
                    f.write_char(ch)?;
                }
                '\u{0000}'..='\u{001F}' => write!(f, "\\u{:04X}", ch as u32)?,
                _ => f.write_char(ch)?,
            }
        }
    } else {
        f.write_str(val)?;
    }
    f.write_char('"')
}

// Catch error return and set error flag
macro_rules! catch {
    ($self:ident, $call:expr) => {{
        if $call.is_err() {
            $self.error = true;
        }
    }};
}

struct Visitor<'a, 'b: 'a> {
    fmt: &'a mut fmt::Formatter<'b>,
    fmtbuf: String,
    prefix: &'static str, // Whatever needs adding before the next item, or ""
    empty: bool,
    error: bool,
}

impl<'a, 'b> Visitor<'a, 'b> {
    fn new(fmt: &'a mut fmt::Formatter<'b>, prefix: &'static str) -> Self {
        Self {
            fmt,
            fmtbuf: String::new(),
            prefix,
            empty: true,
            error: false,
        }
    }
    fn push_key(&mut self, key: Option<&str>) {
        catch!(self, self.fmt.write_str(self.prefix));
        self.prefix = ",";
        self.empty = false;
        if let Some(key) = key {
            catch!(self, push_str_literal(self.fmt, key));
            catch!(self, self.fmt.write_char(':'));
        }
    }
}

impl<'a, 'b> LogVisitor for Visitor<'a, 'b> {
    fn kv_u64(&mut self, key: Option<&str>, val: u64) {
        self.push_key(key);
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_i64(&mut self, key: Option<&str>, val: i64) {
        self.push_key(key);
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_f64(&mut self, key: Option<&str>, val: f64) {
        self.push_key(key);
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_bool(&mut self, key: Option<&str>, val: bool) {
        self.push_key(key);
        catch!(self, write!(self.fmt, "{}", val));
    }
    fn kv_null(&mut self, key: Option<&str>) {
        self.push_key(key);
        catch!(self, self.fmt.write_str("null"));
    }
    fn kv_str(&mut self, key: Option<&str>, val: &str) {
        self.push_key(key);
        catch!(self, push_str_literal(self.fmt, val));
    }
    fn kv_fmt(&mut self, key: Option<&str>, val: &Arguments<'_>) {
        self.push_key(key);
        if self.fmtbuf.capacity() == 0 {
            self.fmtbuf = String::with_capacity(1024);
        }
        self.fmtbuf.clear();
        catch!(self, write!(self.fmtbuf, "{}", val));
        catch!(self, push_str_literal(self.fmt, &self.fmtbuf));
    }
    fn kv_map(&mut self, key: Option<&str>) {
        self.push_key(key);
        catch!(self, self.fmt.write_char('{'));
        self.prefix = "";
    }
    fn kv_mapend(&mut self, _: Option<&str>) {
        catch!(self, self.fmt.write_char('}'));
        self.prefix = ",";
    }
    fn kv_arr(&mut self, key: Option<&str>) {
        self.push_key(key);
        catch!(self, self.fmt.write_char('['));
        self.prefix = "";
    }
    fn kv_arrend(&mut self, _: Option<&str>) {
        catch!(self, self.fmt.write_char(']'));
        self.prefix = ",";
    }
}

#[cfg(test)]
mod test {
    use super::{KvToJson, LogVisitor};
    use std::fmt::Write;

    fn kvscan_empty(_lv: &mut dyn LogVisitor) {}

    fn kvscan_simple(lv: &mut dyn LogVisitor) {
        lv.kv_u64(Some("u64"), 123456789);
        lv.kv_str(Some("str"), "ABCDEFGHIJ");
    }

    fn kvscan_all(lv: &mut dyn LogVisitor) {
        lv.kv_u64(Some("u64"), 123456789);
        lv.kv_i64(Some("i64"), -123456789);
        lv.kv_f64(Some("f64"), 12345.6789);
        lv.kv_bool(Some("b0"), false);
        lv.kv_bool(Some("b1"), true);
        lv.kv_null(Some("null"));
        lv.kv_str(Some("str"), "ABCDEFGHIJ");
        lv.kv_str(Some("str_ctrl"), "ABC\tDEF");
        lv.kv_str(Some("str_quote"), "ABC\"DEF\"GHI");
        lv.kv_str(Some("str_bsl"), "ABC\\DEF\\GHI");
        lv.kv_fmt(Some("fmt"), &format_args!("{}{}{}", "ABC", 123, "DEF"));
        lv.kv_map(Some("map"));
        lv.kv_u64(Some("map_u64"), 987654321);
        lv.kv_str(Some("map_str"), "JIHGFEDCBA");
        lv.kv_map(Some("map_nested"));
        lv.kv_bool(Some("map_nested_bool"), false);
        lv.kv_mapend(Some("map_nested"));
        lv.kv_mapend(Some("map"));
        lv.kv_map(Some("map_empty"));
        lv.kv_mapend(Some("map_empty"));
        lv.kv_arr(Some("arr"));
        lv.kv_u64(None, 987654321);
        lv.kv_str(None, "JIHGFEDCBA");
        lv.kv_arr(None);
        lv.kv_bool(None, true);
        lv.kv_arrend(None);
        lv.kv_arrend(Some("arr"));
        lv.kv_arr(Some("arr_empty"));
        lv.kv_arrend(Some("arr_empty"));
    }

    fn append(
        s: &mut String,
        kvscan: &dyn Fn(&mut dyn LogVisitor),
        prefix: &'static str,
        suffix: &'static str,
    ) {
        write!(s, "{}", KvToJson::new(kvscan, prefix, suffix)).unwrap();
    }

    #[test]
    fn test() {
        // To verify JSON with `jq . -` (for example), run `cargo test
        // -- --nocapture` and paste in the printed JSON data
        let mut buf = "{\"dummy\":1".to_string();
        append(&mut buf, &kvscan_empty, ",", "");
        buf.push_str("}");
        println!("{}", buf);
        assert_eq!(buf, "{\"dummy\":1}");

        let mut buf = "{\"dummy\":1".to_string();
        append(&mut buf, &kvscan_simple, ",", "");
        buf.push_str("}");
        println!("{}", buf);
        assert_eq!(
            buf,
            "{\"dummy\":1,\"u64\":123456789,\"str\":\"ABCDEFGHIJ\"}"
        );

        let mut buf = "{\"dummy\":1".to_string();
        append(&mut buf, &kvscan_empty, ",\"kv\":{", "}");
        buf.push_str("}");
        println!("{}", buf);
        assert_eq!(buf, "{\"dummy\":1}");

        let mut buf = "{\"dummy\":1".to_string();
        append(&mut buf, &kvscan_simple, ",\"kv\":{", "}");
        buf.push_str("}");
        println!("{}", buf);
        assert_eq!(
            buf,
            "{\"dummy\":1,\"kv\":{\"u64\":123456789,\"str\":\"ABCDEFGHIJ\"}}"
        );

        let mut buf = "{\"dummy\":1".to_string();
        append(&mut buf, &kvscan_all, ",", "");
        buf.push_str("}");
        println!("{}", buf);
        assert_eq!(
            buf,
            "{\"dummy\":1,\"u64\":123456789,\"i64\":-123456789,\"f64\":12345.6789,\"b0\":false,\"b1\":true,\"null\":null,\"str\":\"ABCDEFGHIJ\",\"str_ctrl\":\"ABC\\u0009DEF\",\"str_quote\":\"ABC\\\"DEF\\\"GHI\",\"str_bsl\":\"ABC\\\\DEF\\\\GHI\",\"fmt\":\"ABC123DEF\",\"map\":{\"map_u64\":987654321,\"map_str\":\"JIHGFEDCBA\",\"map_nested\":{\"map_nested_bool\":false}},\"map_empty\":{},\"arr\":[987654321,\"JIHGFEDCBA\",[true]],\"arr_empty\":[]}"
        );
    }
}
